src/common/utils/downloadFile.ts

```diff
@@ -0,0 +1,36 @@
+import request from "request";
+
+export interface DownloadFileOptions {
+  url: string;
+  gzip?: boolean;
+}
+
+export interface DownloadFileTicket {
+  url: string;
+  promise: Promise<Buffer>;
+  cancel(): void;
+}
+
+export function downloadFile(opts: DownloadFileOptions): DownloadFileTicket {
+  const { url, gzip = true } = opts;
+  const fileChunks: Buffer[] = [];
+  const req = request(url, { gzip });
+  const promise: Promise<Buffer> = new Promise((resolve, reject) => {
+    req.on("data", (chunk: Buffer) => {
+      fileChunks.push(chunk);
+    });
+    req.on("complete", () => {
```

Comment:
complete and error events should be reacted with ".once"

Reply:
What's the difference in that case?

Reply:
Promise should be resolved or rejected once, not multiple times. The events might only be emitted once in reality, but it's better/safer to only react to the first event since we want the callback to only react the first time.

Reply:
Ok, but can't get why it's safer when:
```
const test = new Promise(resolve => {
  resolve(1);
  setTimeout(() => {
    resolve(2);
    console.log("resolving again won't trigger any errors");
  }, 1000)
});
```

Reply:
Resolving promise twice is an "error" and can lead to difficult to understand/debug bugs, even if there's no error logged for it.